ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTestUtil.java

```diff
@@ -0,0 +1,535 @@
+        q -> queries.addAll(ksqlEngine.buildMultipleQueries(q, Collections.emptyMap()))
+    );
+    return new TopologyTestDriver(queries.get(queries.size() - 1).getTopology(),
+        streamsProperties,
+        0);
+  }
+
+  static List<String> findTests(final String dir) throws IOException {
+    final List<String> tests = new ArrayList<>();
+    try (final BufferedReader reader =
+             new BufferedReader(
+                 new InputStreamReader(EndToEndEngineTestUtil.class.getClassLoader().
+                     getResourceAsStream(dir)))) {
+
+      String test;
+      while ((test = reader.readLine()) != null) {
+        if (test.endsWith(".json")) {
+          tests.add(test);
+        }
+      }
+    }
+    return tests;
+  }
+
+  static void shouldBuildAndExecuteQuery(final Query query) {
+    final MetaStore metaStore = new MetaStoreImpl(new InternalFunctionRegistry());
+    final Map<String, Object> config = new HashMap<String, Object>() {{
+      put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:0");
+      put("application.id", "KSQL-TEST");
+      put("commit.interval.ms", 0);
+      put("cache.max.bytes.buffering", 0);
+      put("auto.offset.reset", "earliest");
+      put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+    }};
+    final Properties streamsProperties = new Properties();
+    final SchemaRegistryClient schemaRegistryClient = new MockSchemaRegistryClient();
+
+    streamsProperties.putAll(config);
+
+    try (final KsqlEngine ksqlEngine = new KsqlEngine(
+        new KsqlConfig(config),
+        new FakeKafkaTopicClient(),
+        schemaRegistryClient,
+        metaStore
+    )) {
+      query.initializeTopics(ksqlEngine);
+      final TopologyTestDriver testDriver = buildStreamsTopology(query, ksqlEngine, streamsProperties);
+      query.processInput(testDriver, schemaRegistryClient);
+      query.verifyOutput(testDriver, schemaRegistryClient);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  static Object valueSpecToAvro(final Object spec, final org.apache.avro.Schema schema) {
+    if (spec == null) {
+      return null;
+    }
+    switch (schema.getType()) {
+      case INT:
+      case LONG:
+      case STRING:
+      case DOUBLE:
+      case FLOAT:
+      case BOOLEAN:
+        return spec;
+      case ARRAY:
+        return ((List)spec).stream()
+            .map(o -> valueSpecToAvro(o, schema.getElementType()))
+            .collect(Collectors.toList());
+      case MAP:
+        return ((Map<Object, Object>)spec).entrySet().stream().collect(
+            Collectors.toMap(
+                Map.Entry::getKey,
+                e -> valueSpecToAvro(e.getValue(), schema.getValueType())
+            )
+        );
+      case RECORD:
+        final GenericRecord record = new GenericData.Record(schema);
+        for (org.apache.avro.Schema.Field field : schema.getFields()) {
+          record.put(field.name(), ((Map<String, ?>)spec).get(field.name()));
```

Comment:
Shouldn't we also call `valueSpecToAvro` on each child field?

Reply:
Good call. I hadn't got to test that case when this was written, since the struct support in KSQL wasn't quite there yet.